Selective IEEE 802.15.4 TX power

jciupis opened the pull request:
This PR adds the possibility to conveniently specify desired TX power for every IEEE 802.15.4 transmission.

Commit:
net: pkt: add IEEE 802.15.4 TX power field

This commit extends the `struct net_pkt` structure with
`ieee802154_txpwr` field that contains signed value of the desired
transmission power of a IEEE 802.15.4 frame in dBm.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

Commit:
net: lib: openthread: use ieee802154_txpwr property

This commit aligns openthread radio platform to use `ieee802154_txpwr`
of the packet to transmit instead of setting power through a separate
API call, if possible.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

ankuns commented on include/net/net_pkt.h:
```diff
@@ -239,7 +239,22 @@ struct net_pkt {
+					  * is valid at the moment.
+					  */
+	};
+	uint8_t ieee802154_txpwr_use : 1; /* Indicates if provided TX power should
```
Position in structure. Could you please check (possibly with godbolt.org) if putting this field here instead of mixed sowewhere with other bitfields just after `ieee802154_mac_hdr_rdy` requires additional allocation unit?
See https://en.cppreference.com/w/c/language/bit_field

jciupis replied:
The structure has multiple variants controlled by Kconfig options, so the answer is not obvious. However, I've checked the part under `#if defined(CONFIG_IEEE802154)` and in some cases rearranging the `ieee802154_txpwr_use` field saves 1B. It's not a staggering saving, but there might be many variables of this type alive at a time, each wasting this one byte.

I'll rearrange the 802.15.4-related fields to shrink the structure's size.

edmont replied:
Could an invalid `ieee802154_txpwr` value be reserved instead of `ieee802154_txpwr_use` to simplify the code?

jciupis replied:
@edmont I've considered using some specific value of `ieee802154_txpwr` to indicate that it should be ignored, yes. However, in my opinion it doesn't simplify the code really. Here goes the explanation:

Let's assume that the nRF IEEE 802.15.4 radio driver exposes a macro, which can be used in `ieee802154_nrf5.c` to indicate that a given transmission should use the default TX power. That macro is only known to `ieee802154_nrf5.c`. Higher layers such as Openthread radio platform implementation are not aware what exact hardware they're being executed on - they call ieee802154 APIs through pointers. So yet another macro would have to be introduced, so that they can be mapped into each other. All other vendors would have to use the same constant too. The flag is easier to pass up and down between layers.

ankuns commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1794,6 +1794,10 @@ static void clone_pkt_attributes(struct net_pkt *pkt, struct net_pkt *clone_pkt)
 
 #if defined(CONFIG_IEEE802154)
 	net_pkt_set_ieee802154_rssi(clone_pkt, net_pkt_ieee802154_rssi(pkt));
+#if defined(CONFIG_IEEE802154_SELECTIVE_TXPOWER)
+	net_pkt_set_ieee802154_txpwr(clone_pkt, net_pkt_ieee802154_txpwr(pkt));
+	net_pkt_set_ieee802154_txpwr_use(clone_pkt, net_pkt_ieee802154_txpwr_use(pkt));
+#endif
```
rssi and txpwr use the same memory in union. This code causes the memory to be copied twice.

jciupis replied:
So you'd suggest to have:
```c
        net_pkt_set_ieee802154_rssi(clone_pkt, net_pkt_ieee802154_rssi(pkt));
#if defined(CONFIG_IEEE802154_SELECTIVE_TXPOWER)
        /* ieee802154_rssi and ieee802154_txpwr form a union, copying one of them is enough */
	net_pkt_set_ieee802154_txpwr_use(clone_pkt, net_pkt_ieee802154_txpwr_use(pkt));
#endif
```

ahasztag commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -298,7 +298,13 @@ void transmit_message(struct k_work *tx_job)
+	net_pkt_set_ieee802154_txpwr(tx_pkt, get_transmit_power_for_channel(channel));
+	net_pkt_set_ieee802154_txpwr_use(tx_pkt, true);
+#else
 	radio_api->set_txpower(radio_dev, get_transmit_power_for_channel(channel));
```
https://github.com/zephyrproject-rtos/zephyr/blob/098602f4aa2460e0b7b61c445b1c5e44a201703b/subsys/net/lib/openthread/platform/radio.c#L648 - I think it should be changed here, too

ahasztag commented:
@jciupis Will this wait until the the PR changing the API will apear in modules/hal_nordic?

jciupis commented:
@ahasztag I can extend the PR with necessary alignments in `drivers/ieee802154/ieee802154_nrf5.c` and a manifest update for https://github.com/zephyrproject-rtos/hal_nordic when you open a PR there. We'll have all needed changes in one place then. What do you think?

Commit:
drivers: ieee802154: nrf: set TX power selectively

This commit extends Nordic's ieee802154 driver with the possibility to
set Tx power for every transmission separately.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

ahasztag commented:
> @ahasztag I can extend the PR with necessary alignments in `drivers/ieee802154/ieee802154_nrf5.c` and a manifest update for https://github.com/zephyrproject-rtos/hal_nordic when you open a PR there. We'll have all needed changes in one place then. What do you think?

There is still some work to do on this PR, I thought there was no sense of opening it until it passes our internal review. Do you want that I open it anyway to be able to reference the new API?

jciupis commented:
@ahasztag no need to rush it :) it would not speed up merge of this PR anyway.

jciupis commented:
CC: @edmont

jciupis commented on west.yml:
```diff
@@ -88,7 +88,7 @@ manifest:
         - hal
     - name: hal_nordic
-      revision: 1f9145e8c8359f103b4ac5085c6f5d2e66a2557d
+      revision: pull/112/head
```
Done

jciupis commented on include/net/net_pkt.h:
```diff
@@ -239,8 +239,31 @@ struct net_pkt {
+	union {
+		uint8_t ieee802154_rssi; /* Received Signal Strength Indication */
+#if defined(CONFIG_IEEE802154_SELECTIVE_TXPOWER)
+		int8_t ieee802154_txpwr; /* TX power. It should be clear from
```
Fixed.

jciupis commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -298,7 +298,13 @@ void transmit_message(struct k_work *tx_job)
+
+#if defined(CONFIG_IEEE802154_SELECTIVE_TXPOWER)
+	net_pkt_set_ieee802154_txpwr(tx_pkt, get_transmit_power_for_channel(channel));
+	net_pkt_set_ieee802154_txpwr_use(tx_pkt, true);
```
@edmont raised the same issue before. The reason I wanted to avoid introducing such a macro was that it cannot be easily passed between generic interfaces connected through pointers to APIs, so very similar macros would have to be created in other parts of the system. However, `struct net_pkt` is used by higher networking layers as well. If the decision about TX power is made when `struct net_pkt` is already available, then we could introduce a macro in `net_pkt.h` such as `NET_PKT_IEEE802154_TXPWR_IGNORE` and use it instead of the flag. I agree that it would be cleaner (no redundancy). @edmont @ahasztag what are your opinions on this?

ahasztag replied:
Wait, my question is different. As I understand when we call `net_pkt_set_ieee802154_txpwr` then we _always_ do it with the intention to use that power - otherwise there is no point of calling this function. No magic value needed.

ahasztag replied:
Note this is different than the case in the radio driver when the value is passed by metadata  there was no separate function for setting the txpwr for a given packet

jciupis replied:
> Wait, my question is different.

Sorry @ahasztag , I haven't read your comment carefully enough. I get that the redundancy is irritating, but I'm wondering if "optimizing" it will not cause more chaos than this additional line of code. Suppose you're preparing a packet and you want to ensure that the default TX power will be used to transmit it. If we remove the setter of `ieee802154_txpwr_use` bit, the only option is to rely on the field to be _already_ set to zero e.g. on initialization. That may not be true, for instance in case of reusing some static `struct net_pkt`.

ahasztag commented:
LGTM

edmont reviewed (changes requested):
Please rebase.

jciupis commented:
@edmont rebased and dropped the manifest update that's no longer necessary.
